ethers-providers/src/pending_escalator.rs

```diff
@@ -0,0 +1,193 @@
+    /// Middleware trait. Callers MUST ensure that transactions are in _reverse_
+    /// broadcast order (this just makes writing the code easier, as we
+    /// can use `pop()` a lot)
+    ///
+    /// TODO: consider deserializing and checking invariants (gas order, etc.)
+    pub(crate) fn new(provider: &'a Provider<P>, mut txns: Vec<Bytes>) -> Self {
+        if txns.is_empty() {
+            panic!("bad args");
+        }
+
+        let first = txns.pop().expect("bad args");
+        // Sane-feeling default intervals
+        Self {
+            provider,
+            broadcast_interval: Duration::from_millis(150),
+            polling_interval: Duration::from_millis(10),
+            txns,
+            last: None,
+            sent: vec![],
+            state: PendingStates::Initial(Box::pin(provider.send_raw_transaction(first))),
+        }
+    }
+
+    pub fn broadcast_interval(mut self, duration: u64) -> Self {
+        self.broadcast_interval = Duration::from_secs(duration);
+        self
+    }
+
+    pub fn polling_interval(mut self, duration: u64) -> Self {
+        self.polling_interval = Duration::from_secs(duration);
+        self
+    }
+}
+
+macro_rules! check_all_receipts {
+    ($cx:ident, $this:ident) => {
+        let futs: Vec<_> = $this
+            .sent
+            .iter()
+            .map(|tx_hash| $this.provider.get_transaction_receipt(*tx_hash))
+            .collect();
+        *$this.state = CheckingReceipts(futs);
+        $cx.waker().wake_by_ref();
+        return Poll::Pending
+    };
+}
+
+macro_rules! sleep {
+    ($cx:ident, $this:ident) => {
+        *$this.state = PendingStates::Sleeping(std::time::Instant::now());
+        $cx.waker().wake_by_ref();
+        return Poll::Pending
+    };
+}
+
+macro_rules! completed {
+    ($this:ident, $output:expr) => {
+        *$this.state = Completed;
+        return Poll::Ready($output)
+    };
+}
+
+macro_rules! broadcast_checks {
+    ($cx:ident, $this:ident, $fut:ident) => {
+        match $fut.as_mut().poll($cx) {
+            Poll::Ready(Ok(pending)) => {
+                $this.sent.push(*pending);
+                check_all_receipts!($cx, $this);
+            }
+            Poll::Ready(Err(e)) => {
+                completed!($this, Err(e));
+            }
+            Poll::Pending => return Poll::Pending,
+        }
+    };
+}
+
+impl<'a, P> Future for EscalatingPending<'a, P>
+where
+    P: JsonRpcClient,
+{
+    type Output = Result<TransactionReceipt, ProviderError>;
+
+    fn poll(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        use PendingStates::*;
+
+        let this = self.project();
+
+        match this.state {
+            Initial(fut) => {
+                broadcast_checks!(cx, this, fut);
+            }
+            Sleeping(instant) => {
+                if instant.elapsed() > *this.polling_interval {
```

Comment:
@gakonst please check this pattern out, as I am unclear why we use `Delay` futures elsewhere. is there some significant behavior difference I'm missing?

Reply:
the futures_timer Delay type uses an additional global timerhandle thread that makes sure the delayed task gets woken up at the right time, however, futures_timer hasn't been updated in a while there's also tokio::time::Sleep now, which I guess works similarly but probably uses some Runtime Handle instead.

looking at the Sleeping state, there might be a case were the Future returns Pending and gets not woken up again (elpased < polling_interval)

Reply:
`tokio::time::Sleep` won't work in WASM iirc

I believe all common executors schedule futures for polling in frames regardless of whether the waker has explicitly requested waking

Reply:
didn't finish the previous thought before submitting the tx

given that this could cause issues with uncommon executors, should we audit all our futures to ensure that all polls either request rewaking _or_ poll another future?